x/archivedb/key.go

```diff
@@ -4,85 +4,70 @@
 package archivedb
 
 import (
+	"bytes"
 	"encoding/binary"
 	"errors"
-	"math"
+
+	"github.com/ava-labs/avalanchego/database"
 )
 
 var (
 	ErrInsufficientLength = errors.New("packer has insufficient length for input")
 	longLen               = 8
-	boolLen               = 1
-	// This default prefix is used as the internal default prefix for all keys.
-	// By doing so the database can safely accept keys from untrusted sources,
-	// by using a common prefix no matter how the keys are crafted externally
-	// they won't interfere with the database
-	//
-	// Having this common prefix also also allows archivedb to store metadata
-	// that may be needed, without polluting the key space.
-	//
-	// Another side effect is that migrations are possible, by changing the
-	// prefix namespace, pulling the old prefix and migrating to the new one.
-	internalKeyPrefix    = []byte("v1")
-	internalKeyPrefixLen = len(internalKeyPrefix)
-	// The length of the suffix data of the key
-	internalKeySuffixLen = longLen + boolLen
+	internalKeySuffixLen  = longLen
+	minInternalKeyLen     = internalKeySuffixLen + 2
 )
 
-// keyInternal
+// dbKey
 //
 // The keys contains a few extra information alongside the given key. The key is
 // what is known outside of this internal scope as the key, but this struct has
-// more information compacted inside the keyInternal to take advantage of natural
+// more information compacted inside the dbKey to take advantage of natural
 // sorting. The inversed height is stored right after the Prefix, which is
-// MaxUint64 minus the desired Height. An extra byte is also packed inside the
-// key to indicate if the key is a deletion of any previous value (because
-// archivedb is an append only database).
+// MaxUint64 minus the desired Height.
 //
-// Any other property are to not serialized but they are useful when parsing a
-// keyInternal struct from the database
-type keyInternal struct {
-	key       []byte
-	height    uint64
-	isDeleted bool
+// All keys are prefixed with a VarUint64, which is the length of the keys, the
+// idea is to make reads even simpler, making it effectively a O(1) operation.
+// Before this change it would have been O(1) unless a another key existed with
+// the same prefix as the requested. By appending the length of the inner key as
+// the first bytes, it is not longer possible to query by a prefix (the full key
+// is needed to do a lookup) but all reads are a O(1).
```

Comment:
I think we can omit how this _used to_ work and just document how it _currently_ works.